### ISSUE TITLE
Add OBFD Packing to TRL/Torch Titan

### ISSUE THREAD

**Issue by abhash-er** [OpenEuroLLM/Taskboard#166]
In [TRL](https://github.com/OpenEuroLLM/post-training), sequence packing currently relies on basic concatenation or simple "Best-Fit Bin Packing." These methods often lead to sub-optimal GPU utilization and complex attention masking requirements.

[OBFD](https://arxiv.org/pdf/2404.10830) (Overlapped Block-wise Fixed-size Dictionary) packing, as implemented in [OLMo-core](https://github.com/allenai/OLMo-core), offers a more deterministic and efficient way to pack variable-length sequences into fixed-size blocks. It would offer us-

- It minimizes padding tokens across distributed ranks, ensuring every training block is "full."
- It improves TFLOPS and training throughput by maximizing compute density.
- OLMo-3 utilizes this strategy to achieve significantly faster SFT speeds.

_Implementation caveats:_

_a) TRL: The target is to pre-pack and save the data as the SFT-Trainer accepts using the pre-tokenized data. But if we want to implement such packing for DPO-Trainer it would be difficult without creating a fork from TRL._ 

_b) TorchTitan- We can do it in a very similar way as pre-packing or also allow an option to dynamically do it within the SFT data loading script. In any case, this also is dependent upon implementing SFT on [TorchTitan](https://github.com/OpenEuroLLM/titan-oellm)._

**Comment by abhash-er** [comment]
Closing Comments: 

> In TRL, We converged to not adding it as it was not giving a lot of speed benefits. 

I implemented the OBFD sequence packing in torch titan [here](https://github.com/OpenEuroLLM/titan-oellm/blob/post-training/titan_oellm/datasets/sft_dataloader.py#L210)! 

1. `Scan lengths` (_`sft_dataloader.py:227-234`_) — iterates the base dataset, recording each sample's length capped at seq_len.
  2. `Sort decreasing` (_`:236-242`_) — sorts doc indices by length, longest first (the "Decreasing" in BFD).
  3. `Best-fit bin assignment` (_`:244-271`_) — for each doc, finds the bin whose remaining capacity is smallest while still fitting it (best-fit); otherwise opens a new bin. Logs fill_ratio.
  4. `__getitem__` (_`:290-318`_) — materializes each bin into a fixed seq_len row (input/labels/loss_mask), pad-filled.

## Usage
Add one line to the [data] section of whichever TOML you run (e.g. titan_oellm/configs/sft_config.toml):
```toml
  [data]
  # ... your existing data_prefix, dataset_split, etc. ...
  packing_strategy = "obfd"
```
  That's the only required change. The field is already wired end to end:
  - Schema: oellm_job_config.py:69 — packing_strategy: str = "none"
  - Dispatch: sft_dataloader.py:1252-1264
  - Build: sft_dataloader.py:1334 → GlobalOBFDPackedDataset

  Notes

  - Accepted values: "none" (pad, default), "obfd", "sequential" (sft_dataloader.py:1253) — anything else raises.
  - Don't combine with prepacked_dataset = true; that force-disables packing since the data is already packed offline (sft_dataloader.py:1258-1263).

## Prepacking

For larger dataset, the process of global packing can be time consuming that unnecessarily holds the GPU. One can also prepack the dataset like this (submit on a cpu node):

```bash
python dataset_packer.py \
    --data-source "allenai/Dolci-Instruct-SFT" \
    --output-dir /path/to/dolci_prepacked_seq16384 \
    --tokenizer-path /path/to/olmo3-7b \
    --seq-len 16384 \
    --instruction-format auto \
    --long-doc-strategy fragment \
    --num-proc 32 \
    --output-format chunked --num-chunks 16
```

During training though, be careful to use prepacked path for the dataset (see this toml example)
```toml
[data]
  data_prefix = "/path/to/dolci_prepacked_seq16384"
  prepacked_dataset = true
  packing_strategy = "none"   # online packing must be off
```

### When to use what
| | Online OBFD | Prepacked |
| :--- | :--- | :--- |
| **config** | `packing_strategy = "obfd"` | `prepacked_dataset = true` |
| **packing cost** | every run, in-process (linear-scan BFD) | once, offline (segment-tree BFD) |
| **startup** | scans/sorts/packs before training | instant mmap |
| **best for** | iterating quickly, smaller data | large data / repeated runs at fixed seq_len |
